LUCENE-10266 Move nearest-neighbor search on points to core

shahrs87 opened the pull request:
<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Lucene:

* https://issues.apache.org/jira/projects/LUCENE

You will need to create an account in Jira in order to create an issue.

The title of the PR should reference the Jira issue number in the form:

* LUCENE-####: <short description of problem or changes>

LUCENE must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

Please provide a short description of the changes you're making with this pull request.

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [ ] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/lucene/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [ ] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [ ] I have added tests for my changes.

shahrs87 committed:
LUCENE-10266 Move nearest-neighbor search on points to core

shahrs87 committed:
LUCENE-10266 Merge with main branch

shahrs87 commented:
@jpountz I have created this PR as per your suggestion in LUCENE-10266 jira. I have made the following assumptions. Please correct me if needed.
1. I have deleted LatLonPointPrototypeQueries class since there is no other sandobx query in that class. Should I keep the empty class ?
2. I see we have FloatPointNearestNeighbor implementation in sandbox which is similar to NearestNeighbor. Do I need to remove FloatPointNearestNeighbor from sandbox and move it to lucene/core ?
3. I have added this change to API Changes section in CHANGES.txt Please correct if it belongs somewhere else.

Thank you.

shahrs87 commented:
@jpountz  @iverase @mikemccand  Can you guys please help me review this patch ? Tagging Ignacio and Michael since you showed interest in the corresponding jira. Thank you.

jpountz commented on lucene/core/src/java/org/apache/lucene/document/LatLonPoint.java:
```diff
@@ -362,4 +377,72 @@ public static Query newDistanceFeatureQuery(
+   *
+   * <p>This is functionally equivalent to running {@link MatchAllDocsQuery} with a {@link
+   * LatLonDocValuesField#newDistanceSort}, but is far more efficient since it takes advantage of
+   * properties the indexed BKD tree. Currently this only works with {@link Lucene90PointsFormat}
```
this point that it only works with Lucene90PointsFormat is no longer true, can you remove it?

shahrs87 replied:
Removed in the latest revision.

jpountz commented on lucene/core/src/java/org/apache/lucene/search/NearestNeighbor.java:
```diff
@@ -31,12 +31,8 @@
- */
-class NearestNeighbor {
+/** KNN search on top of 2D lat/lon indexed points. */
+public class NearestNeighbor {
```
Can we keep it pkg-private if we have it in oal.document? This is especially important as we're promoting this class to core.

shahrs87 replied:
Yes. Changed in the latest revision.

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10266

shahrs87 committed:
@LUCENE-10266 Addressing review comments

shahrs87 commented:
@jpountz  Thank you for the feedback. I have addressed your comments in the latest revision. I have one question. To make changes in the CHANGES.txt file, will this change go in `API changes` section or `Other`. Please advise.

shahrs87 committed:
LUCENE-10266 Changes.txt

jpountz commented on lucene/core/src/java/org/apache/lucene/document/LatLonPoint.java:
```diff
@@ -362,4 +375,71 @@ public static Query newDistanceFeatureQuery(
+   * @return TopFieldDocs containing documents ordered by distance, where the field value for each
+   *     {@link FieldDoc} is the distance in meters
+   * @throws IllegalArgumentException if the underlying PointValues is not a {@code
+   *     Lucene60PointsReader} (this is a current limitation), or if {@code field} or {@code
```
we removed this limitation that `Lucene60PointsReader` is required

jpountz commented on lucene/core/src/java/org/apache/lucene/document/NearestNeighbor.java:
```diff
@@ -220,7 +216,7 @@ public Relation compare(byte[] minPackedValue, byte[] maxPackedValue) {
 
   /** Holds one hit from {@link NearestNeighbor#nearest} */
-  static class NearestHit {
+  public static class NearestHit {
```
can it remain pkg-private too?

jpountz reviewed (commented):
Thanks, it looks almost ready. Maybe as a next step we could try to make this functionality correct on multi-valued fields.

shahrs87 committed:
Merge branch 'main' of https://github.com/apache/lucene into LUCENE-10266

shahrs87 committed:
Addressing review comments

shahrs87 committed:
Adding tidy

shahrs87 commented:
@jpountz Addressed all of your previous comments. Please review again. Thank you.

jpountz commented:
Thanks @shahrs87 !
